src/Horarium.InMemory/InMemoryRepository.cs

```diff
@@ -0,0 +1,156 @@
+
+            void Update(JobDb x)
+            {
+                x.Status = JobStatus.Executing;
+                x.ExecutedMachine = machineName;
+                x.StartedExecuting = DateTime.UtcNow;
+                x.CountStarted++;
+            }
+
+            return Task.FromResult(UpdateOneAtomic(Filter, Update));
+        }
+
+        public Task AddJob(JobDb job)
+        {
+            var wrappedJob = JobInMemoryWrapper.CreateJobInMemoryWrapper(job);
+            _jobsStorage.AddOrUpdate(job.JobId, wrappedJob, (_, __) => wrappedJob);
+
+            return Task.CompletedTask;
+        }
+
+        public Task FailedJob(string jobId, Exception error)
+        {
+            if (!_jobsStorage.TryGetValue(jobId, out var jobWrapper))
+                return Task.CompletedTask;
+
+            lock (jobWrapper.SyncRoot)
+            {
+                jobWrapper.Job.Status = JobStatus.Failed;
+                jobWrapper.Job.Error = error.Message + error.StackTrace;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task RemoveJob(string jobId)
+        {
+            _jobsStorage.TryRemove(jobId, out var _);
+
+            return Task.CompletedTask;
+        }
+
+        public Task RepeatJob(string jobId, DateTime startAt, Exception error)
+        {
+            if (!_jobsStorage.TryGetValue(jobId, out var jobWrapper))
+                return Task.CompletedTask;
+
+            lock (jobWrapper.SyncRoot)
+            {
+                jobWrapper.Job.Status = JobStatus.RepeatJob;
+                jobWrapper.Job.StartAt = startAt;
+                jobWrapper.Job.Error = error.Message + error.StackTrace;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task AddRecurrentJob(JobDb job)
+        {
+            bool Filter(JobDb x) =>
+                x.JobKey == job.JobKey &&
+                (x.Status == JobStatus.Executing || x.Status == JobStatus.Ready);
+
+            void Update(JobDb x)
+            {
+                x.Cron = job.Cron;
+                x.StartAt = job.StartAt;
+            }
+
+            var result = UpdateOneAtomic(Filter, Update);
+            if (result != null) return Task.CompletedTask;
+
+            var copy = JobInMemoryWrapper.CopyJob(job);
+            Update(copy);
+            
+            return AddJob(copy);
+        }
+
+        public Task AddRecurrentJobSettings(RecurrentJobSettings settings)
+        {
+            _recurrentJobSettingsStorage.AddOrUpdate(settings.JobKey, settings, 
+                (_, __) => settings);
+            
+            return Task.CompletedTask;
+        }
+
+        public Task<string> GetCronForRecurrentJob(string jobKey)
+        {
+            if (!_recurrentJobSettingsStorage.TryGetValue(jobKey, out var settings))
+                throw new Exception("Не найдены настройки для рекуррентного джоба");
```

Comment:
Russian

Reply:
Guess i'll fix it in MongoRepository too then